File: mergesort.py
def merge_sort(arr):
  if len(arr) > 1:
    mid = len(arr) // 2
    left = arr[:mid]
    right = arr[mid:]

    merge_sort(left)
    merge_sort(right)

    merged_i = 0
    left_i = 0
    right_i = 0

    while left_i < len(left) and right_i < len(right):
      if left[left_i] < right[right_i]:
        arr[merged_i] = left[left_i]
        left_i += 1
      else:
        arr[merged_i] = right[right_i]
        right_i += 1
      merged_i += 1

    while left_i < len(left):
      arr[merged_i] = left[left_i]
      left_i += 1
      merged_i += 1

    while right_i < len(right):
      arr[merged_i] = right[right_i]
      right_i += 1
      merged_i += 1

arr = [5,2,1,7,5,8]
merge_sort(arr)
print arr

File: mergesort.js
var merge = function(left, right) {
    var merged = [];
    while(left.length && right.length){
        if(left[0] <= right[0]){
            merged.push(left.shift());
        }else{
            merged.push(right.shift());
        }   
    }
    while(left.length){
        merged.push(left.shift());
    }
    while(right.length){
        merged.push(right.shift());
    }
    return merged;
};

var mergeSort = function(arr) {

    if(arr.length < 2) return arr;

    var middle = arr.length>>1;
    var left = arr.slice(0, middle);
    var right = arr.slice(middle, arr.length);

    return merge(mergeSort(left), mergeSort(right));
};

mergeSort([99, 2, 5, 3])